File: src/dotnet/HQ.Abstractions/ChargeCodes/GetChargeCodesV1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using HQ.Abstractions.Common;
using HQ.Abstractions.Enumerations;

namespace HQ.Abstractions.ChargeCodes
{
    public class GetChargeCodesV1
    {
        public class Request : PagedRequestV1
        {
            public string? Search { get; set; }
            public Guid? Id { get; set; }
            public Guid? ProjectId { get; set; }
            public Guid? ClientId { get; set; }

            public SortColumn SortBy { get; set; } = SortColumn.Code;
            public SortDirection SortDirection { get; set; } = SortDirection.Asc;
            public bool? Billable { get; set; }
            public bool? Active { get; set; }
            public Guid? StaffId { get; set; }
        }

        public enum SortColumn
        {
            Code = 1,
            Billable = 2,
            Active = 3,
            ProjectName = 4,
            QuoteName = 5,
            ServiceAgreementName = 6,
            IsProjectMember = 7
        }

        public class Response : PagedResponseV1<Record>;
        public class Record
        {
            public Guid Id { get; set; }
            public string Code { get; set; } = null!;
            public ChargeCodeActivity Activity { get; set; }
            public bool Billable { get; set; }
            public bool Active { get; set; }
            public string? ProjectName { get; set; }
            public string? QuoteName { get; set; }
            public decimal MaximumTimeEntryHours { get; set; }
            public string? ServiceAgreementName { get; set; }
            public Guid? ProjectId { get; set; }
            public Guid? QuoteId { get; set; }
            public Guid? ServiceAgreementId { get; set; }
            public string? ClientName { get; set; }
            public Guid? ClientId { get; set; }

            public string? Description { get; set; }
            public bool? IsProjectMember { get; set; }
            public int IsProjectMemberSort { get; set; }
        }
    }
}

File: src/angular/hq/src/app/models/charge-codes/get-chargecodes-v1.ts
import { ChargeCodeActivity } from '../../enums/charge-code-activity';
import { PagedRequestV1 } from '../common/paged-request-v1';
import { PagedResponseV1 } from '../common/paged-response-v1';
import { SortDirection } from '../common/sort-direction';

// Interface for the paged request with sorting and optional filters specific to charge codes
export interface GetChargeCodesRequestV1 extends PagedRequestV1 {
  search?: string | null;
  id?: string;
  sortBy: SortColumn;
  sortDirection: SortDirection;
  projectId?: string | null;
  clientId?: string | null;
  billable?: boolean;
  active?: boolean;
  staffId?: string | null;
}

// Enum for specifying the sortable columns in charge code requests
export enum SortColumn {
  Code = 1,
  Billable = 2,
  Active = 3,
  ProjectName = 4,
  QuoteName = 5,
  ServiceAgreementName = 6,
  IsProjectMember = 7,
}

// Interface representing a single charge code record
export interface GetChargeCodeRecordV1 {
  activity: ChargeCodeActivity;
  activities: Activity[];
  id: string;
  code: string;
  billable: boolean;
  active: boolean;
  projectName?: string;
  quoteName?: string;
  serviceAgreementName?: string;
  description?: string;
  quoteId?: string;
  projectId?: string;
  clientId?: string;
  clientName?: string;
  maximumTimeEntryHours: number;
  serviceAgreementId?: string;
  isProjectMember: boolean | null;
  requireTask: boolean;
}
export interface Activity {
  id: string;
  name: string;
}

// Interface for the response containing a list of charge code records
export interface GetChargeCodesResponseV1
  extends PagedResponseV1<GetChargeCodeRecordV1> {}
